<a href="https://colab.research.google.com/github/santiagorivas0/Prediccion-de-Abandono-de-Clientes-Bancarios/blob/main/CrossValidation_RivasQuinteros.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#LIBRERIAS

#MANIPULACION DE DATOS, CALCULOS Y GRÁFICOS
import pandas as pd
import numpy as np
import seaborn as sns 
import matplotlib as mpl
import matplotlib.pyplot as plt
from plotly.subplots import make_subplots
import plotly.graph_objects as go

#WARNINGS
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)


**IMPORTAMOS LOS DATOS**

In [ ]:
df = pd.read_csv('Abandono_Clientes_Bancarios.csv')
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

**DATA WRANGLING** 

In [ ]:
df1 = df.drop(["Surname", "CustomerId","Tenure","RowNumber"], axis=1)

In [ ]:
dummy_gender = pd.get_dummies(df["Gender"], prefix="Gender")
dummy_geography = pd.get_dummies(df["Geography"], prefix="Geography")


In [ ]:
df = pd.concat([df, dummy_gender, dummy_geography], axis=1)

BASE PARA MODELIZAR

In [ ]:
df3 = df[['CreditScore','Age','Balance','NumOfProducts','HasCrCard','IsActiveMember','EstimatedSalary','Gender_Male','Geography_France','Geography_Spain','Exited']]
df3.head()

CreditScore  Age    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0          619   42       0.00              1          1               1   
1          608   41   83807.86              1          0               1   
2          502   42  159660.80              3          1               0   
3          699   39       0.00              2          0               0   
4          850   43  125510.82              1          1               1   

   EstimatedSalary  Gender_Male  Geography_France  Geography_Spain  Exited  
0        101348.88            0                 1                0       1  
1        112542.58            0                 0                1       0  
2        113931.57            0                 1                0       1  
3         93826.63            0                 1                0       0  
4         79084.10            0                 0                1       0

In [ ]:
X = df3.drop('Exited', axis=1)
y = df3['Exited']

**CROSS VALIDATION (K-FOLD)**


In [ ]:
# LIBRERIAS
from sklearn.model_selection import StratifiedKFold, cross_val_score, train_test_split
from sklearn.metrics import accuracy_score
from sklearn.model_selection import KFold
from numpy import mean
from numpy import std
from pandas import read_csv
from sklearn.model_selection import LeaveOneOut
from sklearn.model_selection import StratifiedKFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import RandomForestClassifier

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X,y,train_size=0.70,test_size=0.30,random_state=0)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
model = DecisionTreeClassifier(max_depth=1, max_features=2,random_state=7)

In [ ]:
CV_scores = cross_val_score(model, X_train, y_train, cv=10)
CV_scores

array([0.81857143, 0.83285714, 0.82      , 0.81571429, 0.81      ,
       0.82285714, 0.81571429, 0.82      , 0.82857143, 0.81428571])

In [ ]:
print("El Cross Validation Score es de %.5f ± %0.2f" % (CV_scores.mean(), CV_scores.std()))

El Cross Validation Score es de 0.81986 ± 0.01


In [ ]:
kfold = StratifiedKFold(n_splits=10,shuffle=True)

model = DecisionTreeClassifier(max_depth=1 , max_features=2 ,random_state=7)
CV_scores = cross_val_score(model, X_train, y_train, cv=kfold)

print("El Cross Validation Score es de %.5f ± %0.2f" % (CV_scores.mean(), CV_scores.std()))

El Cross Validation Score es de 0.81986 ± 0.01


**¿TENEMOS UN MODELO UNDERFITING U OVERFITTING?**

In [ ]:

from colorama import Fore

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
def print_scores(model ,X_train , y_train, predictions , cv_splites=10):
    
    print(Fore.BLUE , "El promedio de score en el train es de %.5f" % model.score(X_train, y_train))


    CV_scores = cross_val_score(model, X_train, y_train, cv=cv_splites)
    
    print(Fore.BLACK ,"Los scores del cross validation son: \n",CV_scores)
    print(Fore.BLACK ,"El Score mínimo es de%.3f" % min(CV_scores))
    print(Fore.BLACK ,"El Score máximo es de %.3f" % max(CV_scores))
    print(Fore.YELLOW ,"El Cross validation Score es de %.5f ± %0.2f" % (CV_scores.mean(), CV_scores.std()))
    print(Fore.RED ,"El test Score es de %.5f " % accuracy_score(y_test,predictions))

In [ ]:
model = DecisionTreeClassifier(max_depth=1 , max_features=2 ,random_state=7)
model.fit(X_train, y_train)

predictions = model.predict(X_test)
print_scores(model, X_train, y_train, predictions)

 El promedio de score en el train es de 0.81986
 Los scores del cross validation son: 
 [0.81857143 0.83285714 0.82       0.81571429 0.81       0.82285714
 0.81571429 0.82       0.82857143 0.81428571]
 El Score mínimo es de0.810
 El Score máximo es de 0.833
 El Cross validation Score es de 0.81986 ± 0.01
 El test Score es de 0.81933 


**CONCLUSIÓN**

Nuestro Bias es de 0.81933. 

Tenemos un modelo que no está en situación de Underfitting ni tampoco en Overfitting ya que tenemos que el promedio de score en el train es 0.81986  y el test score es de 0.81933. Finalmente son muy similares. 
